_posts/2020-11-03-torch-tabular/torch_tabular.Rmd

```diff
@@ -0,0 +1,445 @@
+| `cap-color, population, habitat, stalk-surface-below-ring, stalk-color-above-ring, spore-print-color`       | 0.9994   |
++-------------------------------------------------------------------------------------------------------------+----------+
+| `cap-color, population, habitat, stalk-surface-below-ring, stalk-color-above-ring, spore-print-color, odor` | 0.9526   |
++-------------------------------------------------------------------------------------------------------------+----------+
+
+Still 95% correct ... While experiments like this are fun, it looks like they can also tell us something serious: Imagine the case of so-called "debiasing" by removing features like race, gender, or income. How many proxy variables may still be left that allow for inferring the masked attributes?
+
+## A look at the hidden representations
+
+Looking at the weight matrix of an embedding module, what we see are the learned representations of a feature's values. The first categorical column was `cap-shape`; let's extract its corresponding embeddings:
+
+```{r}
+embedding_weights <- vector(mode = "list")
+for (i in 1: length(model$embedder$embeddings)) {
+  embedding_weights[[i]] <- model$embedder$embeddings[[i]]$parameters$weight$to(device = "cpu")
+}
+
+cap_shape_repr <- embedding_weights[[1]]
+cap_shape_repr
+```
+
+    torch_tensor
+    -0.0025 -0.1271  1.8077
+    -0.2367 -2.6165 -0.3363
+    -0.5264 -0.9455 -0.6702
+     0.3057 -1.8139  0.3762
+    -0.8583 -0.7752  1.0954
+     0.2740 -0.7513  0.4879
+    [ CPUFloatType{6,3} ]
+
+The number of columns is three, since that's what we chose when creating the embedding layer. The number of rows is six, matching the number of available categories. We may look up per-feature categories in the dataset description (*agaricus-lepiota.names*):
+
+```{r}
+cap_shapes <- c("bell", "conical", "convex", "flat", "knobbed", "sunken")
+
+```
+
+For visualization, it's convenient to do principal components analysis (but there are other options, like t-SNE). Here are the six cap shapes in two-dimensional space:
+
+```{r}
+
+pca <- prcomp(cap_shape_repr, center = TRUE, scale. = TRUE, rank = 2)$x[, c("PC1", "PC2")]
+
+pca %>%
+  as.data.frame() %>%
+  mutate(class = cap_shapes) %>%
+  ggplot(aes(x = PC1, y = PC2)) +
+  geom_point() +
+  geom_label_repel(aes(label = class)) + 
+  coord_cartesian(xlim = c(-2, 2), ylim = c(-2, 2)) +
+  theme(aspect.ratio = 1) +
+  theme_classic()
+```
+
+```{r, eval = TRUE, echo = FALSE, layout="l-body-outset"}
+knitr::include_graphics("images/cap-shape.png")
+```
+
+Naturally, how interesting you find the results depends on how much you care about the hidden representation of a variable. Naturally as well, analyses like these may quickly turn into an activity where extreme caution is to be applied, as any biases in the data will immediately translate into biased representations. Moreover, reduction to two-dimensional space may or may not be adequate.
```

Comment:
```suggestion
Naturally, how interesting you find the results depends on how much you care about the hidden representation of a variable. Analyses like these may quickly turn into an activity where extreme caution is to be applied, as any biases in the data will immediately translate into biased representations. Moreover, reduction to two-dimensional space may or may not be adequate.
```

Reply:
I don't think the repeated naturally really adds anything. 🙂

Reply:
oh, oh, tried to sneak in a little part of rhetoric but didn't work out ;-)))